### ISSUE TITLE
How to extract to memory?

### ISSUE THREAD

**Issue by DanielSWolf** [npm/node-tar#181]
I have a `.tar.gz` archive from which I want to extract a single file. I'd like this file returned as a `Buffer`, so that I can directly work with its content.

I read the documentation, but only found how to extract files *to the file system*, not to an in-memory `Buffer`. What's the correct way to do that? It would seem rather hacky to have `node-tar` write the file to a temporary directory, then read it back and delete it.

**Comment by matutter** [comment]
AFIAK there is no way. The `Writable` steam created for each entry is created in `unpack.js:~252` cannot be replaced with a user-defined writable.

I was also looking for this functionality as well - it would be really useful to pass a writable stream as the return value to the `filter` method when extracting.

It could look something like....

```
let mystream = // writable stream to concat each _write together
await tar.x({
  file: 'my.tar.gz',
  filter: (path, entry) => {
    return mystream 
  }
}, ['file.txt'])
```

**Comment by DanielSWolf** [comment]
Wow, that's an interesting hack. I'm not sure though whether `/dev/null` works cross-platform. So I'd have to detect the platform and [use a different path on Windows](https://stackoverflow.com/questions/313111/is-there-a-dev-null-on-windows).

Meanwhile, I found a package called [`untar-to-memory`](https://www.npmjs.com/package/untar-to-memory) that has exactly the API I was hoping to find in `tar`. Under the hood, it seems to use the low-level `Parse` API offered by`tar`.

**Comment by rmg** [comment]
The docs could probably use an example for this, like there used to be in [tar@2](https://github.com/npm/node-tar/blob/v2.2.1/examples/reader.js) days.

It would probably be something along these lines, though probably updated to use Promises since I haven't done this sort of thing since node 0.10.x days..

```js
const fs = require('fs');
const tar = require('tar');

// concat-stream, Buffer.concat(), etc.. bl is normally what I reach for.
const file = require('pipe-aware-accumulator-here')();

const tgz = fs.createReadStream('sample.tar.gz');
const parser = new tar.Parse({
    filter: path => path === 'the/one/interesting/file',
    onentry: entry => entry.pipe(file),
});

tgz.pipe(parser);
```

**Comment by DanielSWolf** [comment]
Thanks for the code! I turned it into a reusable TypeScript function. Maybe this can help others with the same problem. It's unit tested, but I'm not sure whether it's the most elegant approach. Feedback is welcome!

```ts
import concatStream = require('concat-stream');
import { Parse as TarParser } from 'tar';
import { PassThrough as PassThroughStream } from 'stream';

function bufferToStream(buffer: Buffer): NodeJS.ReadableStream {
  const stream = new PassThroughStream();
  stream.end(buffer);
  return stream;
}

export function extractFileFromTarball(tarball: Buffer, filePath: string): Promise<Buffer> {
  return new Promise<Buffer>((resolve, reject) => {
    let success = false;
    const parser = new TarParser({
      strict: true,
      filter: (currentPath: string) => {
        const match = currentPath === filePath;
        if (match) success = true;
        return match;
      },
      onentry: (entry: NodeJS.ReadableStream) => entry.pipe(concatStream(resolve)),
    });
    bufferToStream(tarball)
      .pipe(parser)
      .on('end', () => {
        if (!success) {
          reject(new Error(`Could not find file '${filePath}' in tarball.`));
        }
      })
      .on('error', reject);
  });
}
```

**Comment by medfreeman** [comment]
Thanks for the inspiration !

Here's an es6 function making this reusable, thanks to @ZJONSSON !
from [ZJONSSON/node-unzipper](https://github.com/ZJONSSON/node-unzipper/blob/affbf89b54b121e85dcd31adf7b1dfde58afebb7/lib/parseOne.js)

*extractSingleFileStream.mjs*
```javascript
import stream from 'stream';

import gunzip from 'gunzip-maybe';
import tar from 'tar-stream';
import duplexer2 from 'duplexer2';

import { bufferStream } from './bufferStream';

const { extract } = tar;

export const extractSingleFileStream = match => {
  const inStream = stream.PassThrough({ objectMode: true });
  const outStream = stream.PassThrough();
  const re = match instanceof RegExp ? match : match && new RegExp(match);
  let found;

  inStream
    .pipe(gunzip())
    .pipe(extract())
    .on('error', function(err) {
      outStream.emit('error', err);
    })
    .on('entry', function(entry, stream, cb) {
      if (found || (re && !re.exec(entry.name))) {
        stream.resume();
        cb();
      } else {
        found = true;
        out.emit('entry', stream);
        stream.on('error', function(e) {
          outStream.emit('error', e);
        });
        stream
          .pipe(outStream)
          .on('error', function(err) {
            cb(err);
          })
          .on('end', function() {
            cb();
          });
      }
    })
    .on('error', Object) // Silence error as its already addressed in transform
    .on('finish', function() {
      if (!found) {
        outStream.emit('error', new Error('PATTERN_NOT_FOUND'));
      } else {
        outStream.end();
      }
    });

  const out = duplexer2(inStream, outStream);
  out.buffer = function() {
    return bufferStream(outStream);
  };

  return out;
};
```

*bufferStream.mjs*
```javascript
import stream from 'stream';
import buffer from 'buffer';

const { Buffer } = buffer;

export const bufferStream = entry => {
  return new Promise(function(resolve, reject) {
    let entryBuffer = Buffer.from('');
    const bufferStream = stream
      .Transform()
      .on('finish', function() {
        resolve(entryBuffer);
      })
      .on('error', reject);

    bufferStream._transform = function(d, e, cb) {
      entryBuffer = Buffer.concat([entryBuffer, d]);
      cb();
    };
    entry.on('error', reject).pipe(bufferStream);
  });
};
```

**Comment by DanielRuf** [comment]
https://www.npmjs.com/package/memfs could also be an option.

**Comment by isaacs** [comment]
You can use `tar.t()` to do this.  (Ie, just list/parse the file, don't extract it to disk.)

Here's an example: https://gist.github.com/isaacs/1bc87e60ed1e578269ab5b76935c3217

**Comment by WinstonN** [comment]
> https://www.npmjs.com/package/memfs could also be an option.

@DanielRuf how? Do you have an example of what you mean?

**Comment by imsnif** [comment]
Just a heads-up: the tar.t option is great but doesn't work well if you want the data synchronously (i.e. can't rely on the end/finish callback). This can be reproduced well with large archives.

There doesn't seem to be any library (that I found, would love to be corrected here) that works for this use case. The very hacky workaround I found for this is changing the `cwd` given to node-tar to a random string and then deleting it when done.

**Comment by isaacs** [comment]
@imsnif You mean like this?

```js
// have a file, read it synchronously and emit events
const entries = []
tar.t({
  file: 'file.tgz',
  sync: true,
  onentry: entry => entries.push(entry.path), // or whatever you have to do with them
})
console.log(entries)

// have a stream, read it as fast as possible
// of course it's only truly sync if the stream itself is sync
const stream = getSyncTarStreamSomehow()
const entriesFromStream = []
stream.pipe(tar.t({
  sync: true,
  onentry: entry => entriesFromStream.push(entry.path),
})
```

**Comment by isaacs** [comment]
If there's something you want to do with tarballs, and this library can't do it (or can't do it well), I definitely want to know :)

**Comment by imsnif** [comment]
Hey @isaacs, I think this is either me not explaining myself well enough, not fully understanding your meaning, or both. :)

My use case: I have a tar archive on my HD. I need to get the contents (in the form of a buffer) from a single file within that archive. I need to do this synchronously.

I can read the tar from the HD synchronously with `fs.readFileSync`, but then I'd like to parse them. I have not found a way to do this with this library. Would very much love to be corrected.

Thanks for chiming in, in any case!

**Comment by imsnif** [comment]
This is what I'm doing now (hacky and a little slow, because we have to write to the disk, but works in a pinch):

```javascript
const uuid = require('uuid').v4
const path = require('path')
const os = require('os')
const fs = require('fs-extra')
const nodeTar = require('tar')

function tarballPathToMemorySync (tarballPath, fileInTar) {
  let entryBuffer = undefined;
  const onentry = entry => {
    if (entry.path === fileInTar) {
      let data = []
      entry.on('data', c => {
        data.push(c)
      })
      entry.on('finish', () => {
        entryBuffer = Buffer.concat(data);
      })
    }
  }

  const tmpDir = path.join(os.tmpdir(), 'temporary-tars', uuid())
  fs.mkdirpSync(tmpDir)

  const extract = nodeTar.x({
    onentry,
    cwd: tmpDir,
    sync: true
  })
  const tarball = fs.readFileSync(tarballPath)
  extract.write(tarball)
  fs.removeSync(tmpDir)
  return entryBuffer
}

const buf = tarballPathToMemorySync('/path/to/my/tarball.tar', 'some/file/in/my/tarball.json')
console.log(buf.toString()) // prints the contents of some/file/in/my/tarball.json

```

**Comment by thuringia** [comment]
@isaacs I think I'm in the same boat as @imsnif and I'm just not getting it :D
My use case is to extract a set of variable length XML-files from a tarball in S3 to S3. I'm trying to make this work in Lamda, which complicates things slightly, but that should be irrelevant.
Using `tar.x` is not an option as the Gzipped archive is about 4GB, but will be much larger soon, with extraction at about 80GB today. So I need to go the Stream route.
Essentially this boils down to piping each `Entry` into `AWS.S3.upload` and waiting for the upload to finish.

I can extract every file but with a length of 0 like this:
```js
s3.getObject({ Bucket: process.env.BUCKET, Key: `node/${FILE_NAME}` })
        .createReadStream()
        .pipe(tar.list())
        .on("entry", async (entry) => {  
            await s3
              .upload({
                Body: entry,
                Bucket: process.env.BUCKET,
                Key: entry.path,
              })
              .promise();
          }
        })
        .on("error", (error) => reject(error))
        .on("close", () => resolve({ statusCode: 200, body: {} }));
    },
```
So an `Entry` is not neither a `ReadableStream` nor a `Buffer`.
This is also not mentioned in the docs meaning I need a different solution.
So I tried attaching a `data` event to `entry` like mentioned above:
```js
s3.getObject({ Bucket: process.env.BUCKET, Key: `node/${FILE_NAME}` })
        .createReadStream()
        .pipe(tar.list())
        .on("entry", async (entry) => {
          const data = [];

            entry.on("data", (c) => {
              data.push(c);
            });
            entry.on("finish", async () => {
              await s3
                .upload({
                  Body: Buffer.concat(data),
                  Bucket: process.env.BUCKET,
                  Key: `node/${date}/${entry.path}`,
                })
                .promise();
            });
          }
        })
        .on("error", (error) => reject(error))
        .on("close", () => resolve({ statusCode: 200, body: {} }));
    },
```
I only get the first member extracted, also with a length of 0, and the Lambda is being killed due to memory consumption.
This means the stream is reading the entry as expected, but is not waiting for the end of each entry.
Some further experimentation with the S3-API suggests that this approach runs into rate-limiting issues.

I got this running using Python with this code (the loop being required as the Python built-in for tarballs is not stream-aware):
```python
with open(f"s3://{get_bucket_name()}/{FILE_NAME}", "rb") as f:
        with tarfile.open(fileobj=f, mode="r|*") as targz:
            while True:
                obj = targz.next()
                if obj is None:
                    break
                
                with open(
                    f"s3://{get_bucket_name()}/{datetime.date.today().isoformat()}/batches/{obj.name}",
                    "wb",
                ) as out:
                    out.write(targz.extractfile(obj).read())
```
This successfully streams through the archive, it's just to slow leading to a timeout.

Basically, I'm looking for the Node version of this.
I'm very convinced I'm just not seeing something here :D

**Comment by fasiha** [comment]
> If there's something you want to do with tarballs, and this library can't do it (or can't do it well), I definitely want to know :)

@isaacs I think I may have found something 🙃? When you use the `tar.t` method to extract a _single_ file from a _large_ tarball (for example, I downloaded the [7.1 GB tar.gz](https://www.ncei.noaa.gov/data/global-historical-climatology-network-daily/archive/) of the Global Historical Climatology Network daily dataset 🤣) to memory, you're scanning the _entire_ tarball, even if you find the file early?

For example, if you download that huge tarball to `~/Downloads` and run the following script with `tar` rpm-installed:
```js
var tar = require('tar');

function extractFileFromTarball(tarpath, filename) {
  var nentries = 0;
  var data = [];
  var onentry = entry => {
    nentries++;
    if (nentries % 1e4 === 0) { console.log('nentries=' + nentries); }
    if (entry.path === filename) { entry.on('data', c => data.push(c)); }
  };
  tar.t({onentry, file: tarpath, sync: true});
  var buf = Buffer.concat(data);
  return buf.toString('ascii');
}
var {env} = require('process');
var res = extractFileFromTarball(env.HOME + '/Downloads/daily-summaries-latest.tar.gz', 'USC00048829.csv');
console.log(res.slice(0, 100))
```
It takes ~2 minutes on my reasonably powerful desktop with a hybrid SSD+spinner disk and prints `nentries=110000` even if I ask it to fetch a different file e.g. `USW00023234.csv`. Which makes sense, `tar.t` is intended to list the contents of the tarball, so it's going to scan the entire file even if it finds the end of the file it's looking for.

I can imagine a few different ways to try to bail early after `tar.t` has finished finding my life but curious if you had a recommendation.

(I also note that `tar xf daily-summaries-latest.tar.gz USC00048829.csv` takes about 1.25 minutes—a bit faster than this node library, but not that much faster. But you did ask if there was anything this library didn't do well with tarballs, which is probably hard to find!)

**Comment by dtinth** [comment]
Here’s how I do it with modern Node.js APIs:

```js
import tar from 'tar'
import stream from 'stream/promises'

await stream.pipeline(
  // Assuming data arrives by stdin
  process.stdin,

  new tar.Parse({
    // Only process file entries, skip non-files (e.g. directories).
    filter: (path, entry) => entry.type === 'File',

    onentry: async (entry) => {
      let size = 0
      // Consume the file.
      // Note that the onentry callback **must** consume the file. Otherwise,
      // `tar.Parse` will not proceed to the next file and execution will stop.
      // If you don’t want to consume the file, skip it by returning `false`
      // in the `filter` callback.
      for await (const data of entry) {
        size += data.length
      }
      console.log(entry.path, size)
    },
  })
)
```

**Comment by valeeum** [comment]
@dtinth 

How do you bubble up a thrown error from the onentry function?   When I throw an error in onentry, I don't get an error in my pipeline catch handler.

**Comment by dtinth** [comment]
@valeeum Sorry, my `onentry` does not throw errors. I currently don’t know what to do in that case as well.
